controller/src/main/java/io/pravega/controller/server/Main.java

```diff
@@ -86,8 +88,28 @@ public static void main(String[] args) {
                     .restServerConfig(Optional.of(restServerConfig))
                     .build();
 
+            Thread.setDefaultUncaughtExceptionHandler((t, e) -> {
+                log.error("Thread {} with stackTrace {} failed with uncaught exception", t.getName(), t.getStackTrace(), e);              
+            });
+
             ControllerServiceMain controllerServiceMain = new ControllerServiceMain(serviceConfig);
             controllerServiceMain.startAsync();
+            Runtime.getRuntime().addShutdownHook(new Thread(() -> {
+                MemoryMXBean memoryMXBean = ManagementFactory.getMemoryMXBean();
+                memoryMXBean.setVerbose(true);
+                log.error("Shutdown hook memory usage dump: Heap memory usage: {}, non heap memory usage {}", memoryMXBean.getHeapMemoryUsage(),
+                        memoryMXBean.getNonHeapMemoryUsage());
+                
+                log.info("Controller service shutting down");
+                controllerServiceMain.stopAsync();
+                controllerServiceMain.awaitTerminated();
```

Comment:
FYI if the Controller will fail (before or during shutdown) this will throw an exception, so the subsequent code won't execute. There is a helper method in `Services` but it converts everything into a Future. An alternative is to put this in a `try-finally` block and move lines 107-110 to the `finally`